### PR TITLE
feat: Handle JSON object for query results panel

### PR DESCRIPTION
**Pull request by gaurpulkit**
### Problem
https://github.com/AltimateAI/vscode-dbt-power-user/issues/1308
https://altimateai.atlassian.net/browse/AI-1944

### Solution
Added a custom perspective viewer plugin to handle json or large strings

### Screenshot/Demo
https://www.loom.com/share/fce536d347e54ad9b1e5556e601e44f4?sid=97de7e23-3e90-44fd-9471-37896cadff99
Screenshot 2024-08-09 at 5 09 29 PM
Screenshot 2024-08-09 at 5 08 57 PM

### How to test

https://github.com/AltimateAI/vscode-dbt-power-user-tests/pull/66

## Checklist

- [x] I have run this code and it appears to resolve the stated issue
- [ ] `README.md` updated and added information about my change

### CODE REVIEW

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx [hunk 1]
```diff
@@ -203,26 +209,55 @@ const PerspectiveViewer = ({
...
         );
       perspectiveViewerRef.current
         ?.delete()
         .catch((err) =>
-          panelLogger.error("error while deleting perspective viewer", err),
+          panelLogger.error("error while deleting perspective viewer", err)
         );
+
+      window.removeEventListener(
+        "json-viewer",
```
Comment (AdiGajbhiye) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx:258]: key is different from `window.addEventListener`

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx [hunk 2]
```diff
@@ -203,26 +209,55 @@ const PerspectiveViewer = ({
...
+      const detail = event.detail as {
+        type: string;
+        message: string;
+        columnName: string;
+      };
+      setDrawerTitle(detail?.columnName);
+      if (detail?.type === "string") {
+        // adding \n after every 45 characters to make it readable
+        setDrawerData(
+          detail?.message.match(/.{1,45}/g)?.join("\n") ?? detail?.message
```
Comment (AdiGajbhiye) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx:231]: probably better to use `word-break: break-word;` in css

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  td: {
+    textContent: string;
+    style: { cursor: string };
+    onclick: { (): void; (): void };
+  },
+  metadata: { value: string; column_header: string }
+) {
+  // If string length is greater than 20, truncate and add ellipsis
+  if (metadata.value.length > 20) {
+    td.textContent = metadata.value.slice(0, 20) + "...";
```
Comment (AdiGajbhiye) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.tsx:32]: maybe try `text-overflow: ellipsis;`

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
+/* eslint-disable */
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:1]: We should fix the TS errors and make this TS file. Keeping it as JS will make the code non typesafe and difficult to make changes in future

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    text.textContent = td.textContent;
+    td.textContent = "";
+
+    // Adding svg icon to the span element
+    const span = document.createElement("span");
+    const img = document.createElement("img");
+
+    // Getting the current theme and setting the icon color
+    const currentTheme = getGlobalTheme();
+    const iconColor = currentTheme === "dark" ? "white" : "black";
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:38]: we should be able to use theme based css values without actual theme and also white and black are not right values and will look odd in different themes

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    img.src = `data:image/svg+xml,%3Csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2030%2030'%20width='12px'%20fill='${iconColor}'%20height='12px'%3E%3Cpath%20d='M%2025.980469%202.9902344%20A%201.0001%201.0001%200%200%200%2025.869141%203%20L%2020%203%20A%201.0001%201.0001%200%201%200%2020%205%20L%2023.585938%205%20L%2013.292969%2015.292969%20A%201.0001%201.0001%200%201%200%2014.707031%2016.707031%20L%2025%206.4140625%20L%2025%2010%20A%201.0001%201.0001%200%201%200%2027%2010%20L%2027%204.1269531%20A%201.0001%201.0001%200%200%200%2025.980469%202.9902344%20z%20M%206%207%20C%204.9069372%207%204%207.9069372%204%209%20L%204%2024%20C%204%2025.093063%204.9069372%2026%206%2026%20L%2021%2026%20C%2022.093063%2026%2023%2025.093063%2023%2024%20L%2023%2014%20L%2023%2011.421875%20L%2021%2013.421875%20L%2021%2016%20L%2021%2024%20L%206%2024%20L%206%209%20L%2014%209%20L%2016%209%20L%2016.578125%209%20L%2018.578125%207%20L%2016%207%20L%2014%207%20L%206%207%20z'%3E%3C/path%3E%3C/svg%3E`;
+
+    // Styling the image element
+    img.style.paddingTop = "5px";
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:42]: these styles could be done via css? ex: perspective.scss?

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    // If string is a valid JSON, add click event to view JSON
+    td.onclick = () => {
+      dispatchCustomEvent(
+        "string-json-viewer",
+        metadata.value,
+        columnName,
+        "json"
+      );
+    };
+  } else if (metadata.value.length > 20) {
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:89]: use early returns

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    appendImage(td);
+    // else if string is greater than 20, add click event to view full string
+    td.onclick = () => {
+      dispatchCustomEvent(
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:93]: it would be more useful if we emit this event as `onCellClick` in perspectiveviewer props, instead of dispatching to window. We can make use the click event for any functionality in future (for ex: copying the cell content). Keeping it in props will be more maintainable

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    const viewer = this.parentElement;
+    const datagrid = this.regular_table;
+    if (this._dirty) {
+      await this.refresh_cache();
+    }
+
+    for (const td of datagrid.querySelectorAll("td")) {
+      const metadata = datagrid.getMeta(td);
+
+      let type;
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:121]: Better to avoid mutable variables. You could extract this logic to separate function and return the type

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+      if (metadata.x >= 0) {
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:122]: can you add comment in code to explain what this is about?

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx [hunk 11]
```diff
@@ -22,6 +23,14 @@ import { executeRequestInAsync } from "@modules/app/requestExecutor";
 import useQueryPanelState from "@modules/queryPanel/useQueryPanelState";
 import { useQueryPanelDispatch } from "@modules/queryPanel/QueryPanelProvider";
 import { setPerspectiveTheme } from "@modules/queryPanel/context/queryPanelSlice";
+import { Drawer, DrawerRef } from "@uicore";
+
+// The global theme is used to store the current theme of the perspective viewer.
+// This is used to update the icon color in the perspective viewer plugin.
+let GLOBAL_THEME = "";
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx:30]: lets discuss to figure out how to use css instead of this

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx [hunk 12]
```diff
@@ -203,33 +216,67 @@ const PerspectiveViewer = ({
...
+    const handleOpenDrawer = (event: CustomEvent) => {
+      drawerRef.current?.open();
+      const detail = event.detail as {
+        type: string;
+        message: string;
+        columnName: string;
+      };
+      setDrawerTitle(detail?.columnName);
+      if (detail?.type === "string") {
+        // adding \n after every 45 characters to make it readable
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx:238]: why specifically 45 characters?

File: webview_panels/src/modules/queryPanel/components/perspective/perspective.module.scss [hunk 13]
```diff
@@ -1,6 +1,7 @@
 .altimatePerspectiveViewer {
   height: 100%;
   font-size: 1.25rem !important;
+  z-index: 20;
```
Comment (saravmajestic) [inline webview_panels/src/modules/queryPanel/components/perspective/perspective.module.scss:4]: why is this needed?

**Review by saravmajestic — CHANGES_REQUESTED** [review]

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js (hunk 4 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:1]: The code for the plugin is a bit involved types-wise. So, Aditya and I decided to use Js as it works correctly. [2b47fab](https://github.com/AltimateAI/vscode-dbt-power-user/pull/1354/commits/2b47fab24d3456f3e2664ba2c179ae2d08168092) has the original file with types if you want to take a shot at it.

**Review by AdiGajbhiye — APPROVED** [review]
lgtm

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js (hunk 5 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:38]: This is the only way I was able to render the svg there like I already showed you. Adding theme based icons is the only way to dynamically change it. white and black works correctly as seen in the screenshots above.

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js (hunk 6 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:42]: Tried it. Didn't work

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js (hunk 8 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/queryPanel/components/perspective/PerspectivePlugins.js:93]: We aren't using that component in the js, so its props don't work there. I'm using events to communicate the drawer opening.

File: webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx (hunk 12 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/queryPanel/components/perspective/PerspectiveViewer.tsx:238]: looks good

File: webview_panels/src/modules/queryPanel/components/perspective/perspective.module.scss (hunk 13 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/queryPanel/components/perspective/perspective.module.scss:4]: Because perspective viewer's configure button shows up on the Drawer.

**Review by saravmajestic — APPROVED** [review]
